src/Bicep.LangServer/Handlers/BicepDocumentFormattingHandler.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using System.Threading;
+using System.Threading.Tasks;
+using Bicep.Core.PrettyPrint;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Utils;
+using OmniSharp.Extensions.LanguageServer.Protocol.Document;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Bicep.Core.PrettyPrint.Options;
+using Microsoft.Extensions.Logging;
+using Bicep.Core.Syntax;
+using Bicep.LanguageServer.Extensions;
+
+namespace Bicep.LanguageServer.Handlers
+{
+    public class BicepDocumentFormattingHandler : DocumentFormattingHandler
+    {
+        private readonly ILogger<BicepDocumentSymbolHandler> logger;
+        private readonly ICompilationManager compilationManager;
+
+        public BicepDocumentFormattingHandler(ILogger<BicepDocumentSymbolHandler> logger, ICompilationManager compilationManager)
+            : base(CreateRegistrationOptions())
+        {
+            this.logger = logger;
+            this.compilationManager = compilationManager;
+        }
+
+        public override Task<TextEditContainer?> Handle(DocumentFormattingParams request, CancellationToken cancellationToken)
+        {
+            CompilationContext? context = this.compilationManager.GetCompilation(request.TextDocument.Uri);
+
+            if (context == null)
+            {
+                // we have not yet compiled this document, which shouldn't really happen
+                this.logger.LogError("Document formatting request arrived before file {Uri} could be compiled.", request.TextDocument.Uri);
+
+                return Task.FromResult<TextEditContainer?>(null);
+            }
+
+            long indentSize = request.Options.TabSize;
+            IndentKindOption indentKindOption = request.Options.InsertSpaces ? IndentKindOption.Space : IndentKindOption.Tab;
+            bool insertFinalNewline = request.Options.ContainsKey("insertFinalNewline") && request.Options.InsertFinalNewline;

```

Comment:
`InsertFinalNewline` won't work until https://github.com/microsoft/vscode-languageserver-node/commit/4d07d4784331208a4c2e5a4b2238d4cdf67b67eb is released.

We should also remove the ContainsKey guard once https://github.com/OmniSharp/csharp-language-server-protocol/issues/411 is fixed.

Reply:
I'll have a hotfix ready for this as 0.18.3, as that's a pretty big break.

Reply:
https://github.com/OmniSharp/csharp-language-server-protocol/pull/412

Reply:
@david-driscoll Thanks for the quick turn around!